Fix dsmr tests broken by Python 3.14.3 asyncio changes

### Description

The following `dsmr` tests are flaky/failing due to the Python 3.14.3 asyncio executor scheduling change (see home-assistant/core#162263 for background).

CPython PR python/cpython#142358 changed how `run_in_executor` delivers results back to the event loop. Results are now set directly on the asyncio future instead of going through `call_soon_threadsafe`, which means `await` returns immediately but other `call_soon_threadsafe` callbacks may not have run yet.

#### Affected tests

**`tests/components/dsmr/test_sensor.py::test_default_setup`  (failed 6/10)**

```
tests/components/dsmr/test_sensor.py:169: in test_default_setup
    assert power_consumption.state == "35.0"
E   AssertionError: assert 'unavailable' == '35.0'
E
E     - 35.0
E     + unavailable
```

**`tests/components/dsmr/test_diagnostics.py::test_diagnostics` (failed 9/10)**

```
tests/components/dsmr/test_diagnostics.py:82: in test_diagnostics
    assert result == snapshot
E   AssertionError: assert [+ received] == [- snapshot]
E       dict({
E     -   'data': dict({
E     +   'data': None,
E     -     'CURRENT_ELECTRICITY_USAGE': dict({
E     -       'unit': 'W',
E     -       'value': 0.0,
E     -     }),...
E
E     ...Full output truncated (13 lines hidden), use '-vv' to show
```

#### What was done

These tests have been temporarily marked with `@pytest.mark.xfail` in PR #169064.

#### What needs to happen

The codeowner should investigate and fix the underlying timing issue in these tests. Common fixes include:
- Adding `await hass.async_block_till_done()` after service calls
- Converting sync entity methods to async
- Restructuring mocks to avoid unintended executor code paths

### Additional context

- Related PR: home-assistant/core#162263
- Disable PR: #169064

--- a/tests/components/dsmr/conftest.py
+++ b/tests/components/dsmr/conftest.py
@@ -25,6 +25,9 @@ def dsmr_connection_fixture() -> Generator[tuple[MagicMock, MagicMock, MagicMock
     transport = MagicMock(spec=asyncio.Transport)
     protocol = MagicMock(spec=DSMRProtocol)
 
+    closed = asyncio.Event()
+    protocol.wait_closed = closed.wait
+
     async def connection_factory(*args, **kwargs):
         """Return mocked out Asyncio classes."""
         return (transport, protocol)
@@ -42,6 +45,7 @@ def dsmr_connection_fixture() -> Generator[tuple[MagicMock, MagicMock, MagicMock
         ),
     ):
         yield (connection_factory, transport, protocol)
+        closed.set()
 
 
 @pytest.fixture
@@ -52,6 +56,9 @@ def rfxtrx_dsmr_connection_fixture() -> Generator[
 
     transport = MagicMock(spec=asyncio.Transport)
     protocol = MagicMock(spec=RFXtrxDSMRProtocol)
+
+    closed = asyncio.Event()
+    protocol.wait_closed = closed.wait
 
     async def connection_factory(*args, **kwargs):
         """Return mocked out Asyncio classes."""
@@ -70,6 +77,7 @@ def rfxtrx_dsmr_connection_fixture() -> Generator[
         ),
     ):
         yield (connection_factory, transport, protocol)
+        closed.set()
 
 
 @pytest.fixture

--- a/tests/components/dsmr/test_diagnostics.py
+++ b/tests/components/dsmr/test_diagnostics.py
@@ -10,7 +10,6 @@ from dsmr_parser.obis_references import (
     GAS_METER_READING,
 )
 from dsmr_parser.objects import CosemObject, MBusObject, Telegram
-import pytest
 from syrupy.assertion import SnapshotAssertion
 
 from homeassistant.core import HomeAssistant
@@ -20,9 +19,6 @@ from tests.components.diagnostics import get_diagnostics_for_config_entry
 from tests.typing import ClientSessionGenerator
 
 
-@pytest.mark.xfail(
-    reason="Flaky due to Python 3.14.3 asyncio changes - see home-assistant/core#162263"
-)
 async def test_diagnostics(
     hass: HomeAssistant,
     hass_client: ClientSessionGenerator,

--- a/tests/components/dsmr/test_sensor.py
+++ b/tests/components/dsmr/test_sensor.py
@@ -51,9 +51,6 @@ from homeassistant.helpers import entity_registry as er
 from tests.common import MockConfigEntry, patch
 
 
-@pytest.mark.xfail(
-    reason="Flaky due to Python 3.14.3 asyncio changes - see home-assistant/core#162263"
-)
 async def test_default_setup(
     hass: HomeAssistant,
     entity_registry: er.EntityRegistry,

